### DISCUSSION TITLE
check_drivesize, perf-config, extra(size)

### DISCUSSION THREAD

**Discussion by petitPeyre** [mickem/nscp#835] [Q&A]
here is the command I use and the corresponding result.
```
check_drivesize show-all drive=C: "detail-syntax=(${drive}\\) size: ${size}, used:${used} (${used_pct}%)" "perf-config=used(unit:GB;suffix:used;prefix:'') used %(ignored:true) extra(size)"
L        cli OK: OK (C:\) size: 63.899GB, used:41.111GB (65%)
L        cli  Performance data: 'C:_size'=68611469312;0;0 'C:used'=41.11051GB;51.11952;57.50946;0;63.89941
```
to get rid of the drive name in the perf data, I add ```"perf-syntax=none"``` to the command
```
check_drivesize show-all drive=C: "detail-syntax=(${drive}\\) size: ${size}, used:${used} (${used_pct}%)" "perf-config=used(unit:GB;suffix:used;prefix:'') used %(ignored:true) extra(size)" "perf-syntax=none"
L        cli OK: OK (C:\) size: 63.899GB, used:41.111GB (65%)
L        cli  Performance data: '_size'=68611469312;0;0 'used'=41.11051GB;51.11952;57.50946;0;63.89941
```
it works for the ```used``` part but there is still an underscore in front of ```size```     
I also tried to have the metrics for ```size``` in the perf data in GB. I tried both ```size(unit:GB)``` ```*(unit:GB)``` in the ```perf-config``` string. it adds ```GB``` after the size but the value stays in bytes.

```
check_drivesize show-all drive=C: "detail-syntax=(${drive}\\) size: ${size}, used:${used} (${used_pct}%)" "perf-config=used(unit:GB;suffix:used;prefix:'') used %(ignored:true) extra(size) *(unit:GB)" "perf-syntax=none"
L        cli OK: OK (C:\) size: 63.899GB, used:41.111GB (65%)
L        cli  Performance data: '_size'=68611469312GB;0;0 'used'=41.11052GB;51.11952;57.50946;0;63.89941
```

any idea to remove the ```_``` and have the size in GB is welcome !

---
I perform the test with ```nscp test``` in NSClient++ 0.5.1.46

**Comment by mickem** [comment]
Sorry for the late reply. Both things you ran into have an explanation, and the second one is a bug that has since been fixed.

**1. The underscore in `_size`**

Every performance label is built as `<prefix><perf-syntax alias><suffix>`. Keywords that come with their own perf shape (`used`, `free`, ...) register a suffix such as `" used"`, which is why you see `C:\ used` by default. A keyword pulled in with `extra(...)` that has no shape of its own gets the default suffix `_<keyword>`, so `extra(size)` gives `<alias>_size`. `perf-syntax=none` only blanked the alias, which is why the `_size` part survived.

The suffix (and prefix) are perf-config keys just like `unit:`, so you can override them per metric. `none` clears one:

```
perf-config=size(suffix:none) extra(size)          -> label is just the alias, e.g. C:
perf-config=size(suffix:' size') extra(size)       -> C: size
```

Note that in current releases `perf-syntax=none` (or an empty perf-syntax) suppresses *all* per-row perfdata (fixed for #681), so the "empty alias" trick no longer applies. Set `perf-syntax` to the text you want as the base name instead, e.g. `perf-syntax=C` or `perf-syntax=${drive}`, and shape the rest with `prefix:`/`suffix:`.

**2. `unit:GB` relabelled `size` without converting it**

That was a real bug. `used`/`free` use the auto-scaling byte generator, which converts the value and bounds into the requested unit; plain byte keywords such as `size` only changed the *label*, so you got `68611469312GB`. Since 0.17.0, `unit:` on a plain byte series converts the value and the warn/crit bounds as well (see the 0.17.0 upgrade notes, "perf-config's unit: now converts plain byte series instead of relabelling them").

Putting it together, on a current release this should give you `C used` and `C size` both in GB:

```
check_drivesize drive=C: "detail-syntax=(${drive}\\) size: ${size}, used: ${used} (${used_pct}%)" "perf-syntax=C" "perf-config=used(unit:GB;suffix:' used') used %(ignored:true) size(unit:GB;suffix:' size') extra(size)"
```

0.5.1.46 is a 2016 build; the current release is 0.21.0 and there is a lot more than this in between: https://github.com/mickem/nscp/releases
